presto-common/src/main/java/com/facebook/presto/common/block/DictionaryBlock.java

```diff
@@ -264,14 +264,28 @@ public long getRegionLogicalSizeInBytes(int positionOffset, int length)
         }
 
         long sizeInBytes = 0;
-        long[] seenSizes = new long[dictionary.getPositionCount()];
-        Arrays.fill(seenSizes, -1L);
-        for (int i = positionOffset; i < positionOffset + length; i++) {
-            int position = getId(i);
-            if (seenSizes[position] < 0) {
-                seenSizes[position] = dictionary.getRegionLogicalSizeInBytes(position, 1);
+        // Dictionary Block may contain large number of keys and small number of indices.
+        // This happens during the file read or write, where the dictionary of the file
+        // is much larger than number of rows read. In these cases, compute the size
+        // in place and do not use the cache code path. If the length is less than keys
+        // the cache is likely to be not used.
+        if (length > dictionary.getPositionCount()) {
+            // Cache code path.
+            long[] seenSizes = new long[dictionary.getPositionCount()];
+            Arrays.fill(seenSizes, -1L);
+            for (int i = positionOffset; i < positionOffset + length; i++) {
+                int position = getId(i);
+                if (seenSizes[position] < 0) {
+                    seenSizes[position] = dictionary.getRegionLogicalSizeInBytes(position, 1);
+                }
+                sizeInBytes += seenSizes[position];
+            }
+        }
+        else {
+            // In-place code path.
+            for (int i = positionOffset; i < positionOffset + length; i++) {
+                sizeInBytes += dictionary.getRegionLogicalSizeInBytes(getId(i), 1);
```

Comment:
`getRegionLogicalSizeInBytes` is a very expensive call. How likely/useful it would be to have a light-weighted hashmap:
```java
int[] seenIds = new int[length];
long[] seenSize = new long[length];
Arrays.fill(seenIds, -1);
for (int i = positionOffset; i < positionOffset + length; i++) {
    int id = getId(i);
    int index = id % length;
    if (seenIds[index] == id) {
        sizeInBytes += seenSize[index];
    }
    else if (seenIds[index] == -1) {
        seenIds[index] = id;
        seenSize[index] = dictionary.getRegionLogicalSizeInBytes(id, 1);
        sizeInBytes += seenSize[index];
    }
    else {
        sizeInBytes += dictionary.getRegionLogicalSizeInBytes(getId(i), 1);
    }
}
```

Frankly, I don't know if this approach is going to be better or worth. Maybe worth profiling?

Reply:
The seenIds, need to be a regular HashMap. The dictionary index could be 1,4,7 with length 3, in which case all of them will collide and produce wrong results. Adding the actual hash map will make any implementation much worser.

I looked at few different implementations of the getRegionLogicalSizeInBytes (VariableWidthBlock is cheap, but MapBlock is costly). So it is very hard to come up with cost model and even profiling.

What I have in mind is, instead of doing the in-place code path when length <= dictionarySize, we can do the in place code path, when length * 2 <= dictionarySize, this assumes that a hit in the cache is twice cheaper than calling the actual method. I am ok changing the condition to length * 5 <= dictionarySize, so the in place code path is triggered when the cache is not going to yield any benefits.

Or other approach we can do is, using a regular hashMap in the inplace code path and let it grow dynamically.

Reply:
Other approach I coded and discarded is, I can make changes to the dictionary writer to not call the getRegionLogicalSizeInBytes and can do inplace evaluation as the use case is well understood. But I believe the fix here is better as it will improve the performance of Presto. I coded up the dictionary Writer fix and discarded it. I will push that change as well, just in case.